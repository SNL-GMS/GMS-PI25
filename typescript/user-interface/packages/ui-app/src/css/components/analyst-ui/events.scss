.events-status-bar {
  display: flex;
  flex: 0 0 auto;
  justify-content: space-between;
  align-items: center;
}

.events-status-bar__visibility-menu-button {
  padding-left: 0px; // Override for default to make flush left
}

.events-status-bar__event-counters {
  display: flex;
  flex: 0 0 auto;
  justify-content: space-between;
  & > * {
    padding-right: 0.5rem; // All direct children have right padding to enforce spacing
  }
}

.events-status-bar__mark-button {
  display: flex;
  flex: 1 0 auto;
  justify-content: flex-end;
}

.events__mark-complete-button {
  height: 20px;
  min-height: 23px !important;
  min-width: 0px !important;
  position: relative;
  top: -1px;
}

.events__conflict-marker {
  height: 36px;
  flex: auto;
  display: flex;
  justify-content: center;
}

.event-panel {
  display: flex;
  height: 100%;
  width: 100%;
  flex-direction: column;
  overflow-y: auto;
  overflow-x: hidden;
}

.event-table {
  height: 100%;
  width: 100%;
}

.events-display-window {
  @include flexColumn;
  width: 100%;
  height: 100%;
  .toolbar {
    margin-bottom: 0.5em;
  }
}

.table-cell .table-cell__value--complete {
  color: $analyst-complete;
}

.ag-theme-dark .ag-row-even.edge-event-row .table-cell {
  background-color: var(--gms-background, $gms-background);
  filter: brightness(0.5);
}

.ag-theme-dark .ag-row-odd.edge-event-row .table-cell {
  background-color: var(--gms-background, $gms-background);
  filter: brightness(0.5);
}

.ag-theme-dark .ag-row-even.open-event-row .table-cell {
  background-color: var(--analyst-open-event, $analyst-open-event);
  .table-cell__value {
    color: var(--open-associated-text-color, $gms-main);
  }
}

.ag-theme-dark .ag-row-odd.open-event-row .table-cell {
  background-color: var(--analyst-open-event, $analyst-open-event);
  .table-cell__value,
  svg {
    color: var(--open-associated-text-color, $gms-main);
  }
}

.bp4-popover2-content-icon-margin-left {
  margin-left: 15px;
}

.bp4-popover2-content-dialog-container-left-right-padding {
  padding-left: 1.5em;
  padding-right: 1.5em;
}

@container (max-width: 900px) {
  .event-panel {
    .counter__count--prefix {
      display: none;
    }
    .counter__entry + .counter__entry {
      margin-left: 1rem;
    }
    .counter__count {
      margin-left: 0.25rem;
    }
  }
}

@container (max-width: 650px) {
  .event-panel {
    .counter__entry + .counter__entry {
      margin-left: 0.5rem;
    }
    .counter__label {
      display: none;
    }
  }
}
