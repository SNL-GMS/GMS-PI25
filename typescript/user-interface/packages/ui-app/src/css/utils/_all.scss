@import './variables.scss';
@import './string-utils.scss';
@import './layout-utils.scss';
@import './number-utils.scss';
@import './visual-utils.scss';
@import './interaction-utils.scss';
@import './table.scss';
