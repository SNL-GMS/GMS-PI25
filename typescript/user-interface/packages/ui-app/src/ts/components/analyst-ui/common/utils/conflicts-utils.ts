import { EventTypes, SignalDetectionTypes } from '@gms/common-model';
import type { AssociationConflictRecord } from '@gms/ui-state';

import type { ConflictedEventInfo, ConflictedSDInfo } from '../tooltips/conflicts-tooltip-content';

/**
 * Given two objects with a `time` property, sort them in chronological order.
 */
export function sortByTimeChronological(a: { time: number }, b: { time: number }) {
  if (a.time < b.time) return -1;
  if (a.time > b.time) return 1;
  return 0;
}

/**
 *
 * @param allSignalDetectionsInInterval
 * @param allEventsInInterval
 * @param allConflicts
 * @param openIntervalName
 * @param currentEventId
 * @returns the array of signal detections with conflicts associated to the input event
 */
export const getConflictedSDsForEvent = (
  allSignalDetectionsInInterval: SignalDetectionTypes.SignalDetection[],
  allEventsInInterval: EventTypes.Event[],
  allConflicts: AssociationConflictRecord,
  openIntervalName: string,
  currentEventId: string
): ConflictedSDInfo[] => {
  // current event of the row
  const currentEvent: EventTypes.Event = allEventsInInterval?.find(ev => ev.id === currentEventId);
  if (!currentEvent) return [];

  let currentEventHypothesis: EventTypes.EventHypothesis = EventTypes.findPreferredEventHypothesisByStage(
    currentEvent,
    openIntervalName
  );
  if (
    currentEventHypothesis === undefined ||
    currentEventHypothesis.locationSolutions === undefined
  ) {
    currentEventHypothesis = EventTypes.findEventHypothesisParent(
      currentEvent,
      currentEventHypothesis
    );
  }

  const currentEventAssociatedSDIds: string[] =
    currentEventHypothesis?.associatedSignalDetectionHypotheses.map(
      sd => sd.id.signalDetectionId
    ) || [];

  // filter the array to only have sd id's that are associated to the current event
  const associatedSDIdsWithConflicts: string[] =
    currentEventAssociatedSDIds.length > 0
      ? Object.keys(allConflicts).filter(sdid => currentEventAssociatedSDIds.includes(sdid))
      : [];

  const sdConflicts =
    associatedSDIdsWithConflicts.length > 0
      ? allSignalDetectionsInInterval
          ?.filter(sd => associatedSDIdsWithConflicts.includes(sd.id)) // only the sd's associated to the current event with conflicts
          .map<ConflictedSDInfo>(sd => {
            const sdHypothesis = SignalDetectionTypes.Util.getCurrentHypothesis(
              sd.signalDetectionHypotheses
            );
            const sdTime: number = SignalDetectionTypes.Util.findArrivalTimeFeatureMeasurementValue(
              sdHypothesis.featureMeasurements
            )?.arrivalTime?.value;
            const sdPhase = SignalDetectionTypes.Util.findPhaseFeatureMeasurementValue(
              sdHypothesis.featureMeasurements
            )?.value;

            return { station: sd.station.name, phase: sdPhase, time: sdTime };
          })
      : [];

  if (sdConflicts?.length > 1) {
    sdConflicts.sort(sortByTimeChronological);
  }
  return sdConflicts;
};

/**
 * @param allEventsInInterval
 * @param allConflicts
 * @param openIntervalName
 * @param currentSDId
 * @returns the array of events with conflicts because the input SD is signal detections with conflicts associated to the input event
 */
export const getConflictedEventsForSD = (
  allEventsInInterval: EventTypes.Event[],
  allConflicts: AssociationConflictRecord,
  openIntervalName: string,
  currentSDId: string
): ConflictedEventInfo[] => {
  const eventConflicts =
    allConflicts[currentSDId]?.eventHypothesisIds
      .map<ConflictedEventInfo>(ev => {
        const currentEvent = allEventsInInterval?.find(evt => evt.id === ev.eventId);
        if (!currentEvent) return undefined;

        let currentEventHypothesis: EventTypes.EventHypothesis = EventTypes.findPreferredEventHypothesisByStage(
          currentEvent,
          openIntervalName
        );

        // If no preferredHypothesis then default to the parent hypothesis
        if (
          currentEventHypothesis === undefined ||
          currentEventHypothesis.locationSolutions === undefined
        ) {
          currentEventHypothesis = EventTypes.findEventHypothesisParent(
            currentEvent,
            currentEventHypothesis
          );
        }

        if (currentEventHypothesis) {
          const locationSolution = EventTypes.findPreferredLocationSolution(
            currentEventHypothesis.id.hypothesisId,
            currentEvent.eventHypotheses
          );
          return { time: locationSolution?.location?.time };
        }
        return undefined;
      })
      .filter(e => e !== undefined) || [];

  if (eventConflicts?.length > 1) {
    eventConflicts.sort(sortByTimeChronological);
  }
  return eventConflicts;
};
