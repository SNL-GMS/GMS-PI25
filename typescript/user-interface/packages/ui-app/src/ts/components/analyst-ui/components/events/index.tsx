import { IanDisplays, toDisplayTitle } from '@gms/common-model/lib/displays/types';
import { nonIdealStateWithSpinner } from '@gms/ui-core-components';
import React from 'react';

import type { EventsComponentProps } from './events-component';

const EventsLazy = React.lazy(async () =>
  import(/* webpackChunkName: 'ui-app-events' */ './events-component').then(module => ({
    default: module.EventsComponent
  }))
);

export function Events(props: EventsComponentProps) {
  return (
    <React.Suspense
      fallback={nonIdealStateWithSpinner(toDisplayTitle(IanDisplays.EVENTS), 'Loading Display')}
    >
      {/* eslint-disable-next-line react/jsx-props-no-spreading */}
      <EventsLazy {...props} />
    </React.Suspense>
  );
}
