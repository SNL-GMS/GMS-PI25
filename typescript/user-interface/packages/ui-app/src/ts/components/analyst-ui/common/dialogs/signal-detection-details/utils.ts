import { FilterTypes, SignalDetectionTypes } from '@gms/common-model';
import {
  DATE_TIME_FORMAT_WITH_FRACTIONAL_SECOND_PRECISION,
  dateToString,
  toDate
} from '@gms/common-util';
import { FormTypes } from '@gms/ui-core-components';
import flatten from 'lodash/flatten';

import { parseWaveformChannelType } from '~analyst-ui/common/utils/signal-detection-util';
import { messageConfig } from '~analyst-ui/config/message-config';
import { formatNumberForDisplayFixedThreeDecimalPlaces } from '~common-ui/common/table-utils';

import type { SignalDetectionHistoryRow } from './types';

/**
 * Generate the table row data for the detection history.
 */
export function generateDetectionHistoryTableRows(
  detection: SignalDetectionTypes.SignalDetection
): SignalDetectionHistoryRow[] {
  const rows = flatten(
    detection.signalDetectionHypotheses
      .map(signalDetectionHypo => {
        const phaseFMValue = SignalDetectionTypes.Util.findPhaseFeatureMeasurementValue(
          signalDetectionHypo?.featureMeasurements
        );
        const arrivalTimeFM = SignalDetectionTypes.Util.findArrivalTimeFeatureMeasurement(
          signalDetectionHypo?.featureMeasurements
        );
        const arrivalTimeFMValue = SignalDetectionTypes.Util.findArrivalTimeFeatureMeasurementValue(
          signalDetectionHypo?.featureMeasurements
        );
        return {
          id: signalDetectionHypo.id.id,
          versionId: signalDetectionHypo.id.id, // TODO: what should the version be? (hypo number?)
          creationTimestamp: -1, // TODO: change when field is available for now -1 translates to 'TBD'
          phase: phaseFMValue.value,
          arrivalTimeMeasurementFeatureType:
            SignalDetectionTypes.FeatureMeasurementType.ARRIVAL_TIME,
          arrivalTimeMeasurementTimestamp: arrivalTimeFMValue.arrivalTime.value,
          arrivalTimeMeasurementUncertaintySec: arrivalTimeFMValue.arrivalTime.standardDeviation,
          channelSegmentType: parseWaveformChannelType(arrivalTimeFM.channel.name),
          authorName: 'TBD',
          deleted: signalDetectionHypo?.deleted ? 'True' : 'False'
        };
      })
      .sort((a, b) => b.arrivalTimeMeasurementTimestamp - a.arrivalTimeMeasurementTimestamp)
  );
  rows[0]['first-in-table'] = true;
  return rows;
}

/**
 * Given a signalDetection, build a list of {@link FormTypes.FormItem}s to be used
 * in the Signal Detection Details menu
 *
 * @param signalDetection Signal Detection to build against
 * @param filterDefinitionByFilterDefinitionUsage Optional filter definitions being used
 * @returns
 */
export function buildSdDetailFormItems(
  signalDetection: SignalDetectionTypes.SignalDetection,
  filterDefinitionByFilterDefinitionUsage?: FilterTypes.FilterDefinitionByFilterDefinitionUsage
): FormTypes.FormItem[] {
  const formItems: FormTypes.FormItem[] = [];
  const sdHypothesis = SignalDetectionTypes.Util.getCurrentHypothesis(
    signalDetection.signalDetectionHypotheses
  );
  const arrivalTimeFM = SignalDetectionTypes.Util.findArrivalTimeFeatureMeasurement(
    sdHypothesis?.featureMeasurements
  );
  const arrivalTimeFMValue = SignalDetectionTypes.Util.findArrivalTimeFeatureMeasurementValue(
    sdHypothesis?.featureMeasurements
  );
  const fmPhase = SignalDetectionTypes.Util.findPhaseFeatureMeasurementValue(
    sdHypothesis?.featureMeasurements
  );

  // Build the arrival uncertainty formatted string
  let arrivalTimeUncertainty = formatNumberForDisplayFixedThreeDecimalPlaces(
    arrivalTimeFMValue.arrivalTime.standardDeviation
  );
  // If a valid fields add 's' seconds unit
  if (arrivalTimeUncertainty !== messageConfig.invalidCellText) {
    arrivalTimeUncertainty = `${arrivalTimeUncertainty}s`;
  }

  formItems.push(
    {
      itemKey: 'Station',
      labelText: 'Station',
      itemType: FormTypes.ItemType.Display,
      displayText: signalDetection.station.name
    },
    {
      itemKey: 'Creation time',
      labelText: 'Creation time',
      itemType: FormTypes.ItemType.Display,
      displayText: 'TBD'
      // displayTextFormat: FormTypes.TextFormats.Time
    },
    {
      itemKey: 'Phase',
      labelText: 'Phase',
      displayText: fmPhase.value ?? 'Unknown',
      itemType: FormTypes.ItemType.Display
    },
    {
      itemKey: 'Detection time',
      labelText: 'Detection time',
      itemType: FormTypes.ItemType.Display,
      className: 'monospace',
      displayText: `${dateToString(
        toDate(arrivalTimeFMValue?.arrivalTime.value),
        DATE_TIME_FORMAT_WITH_FRACTIONAL_SECOND_PRECISION
      )} Â± ${arrivalTimeUncertainty}`,
      displayTextFormat: FormTypes.TextFormats.Time
    },
    {
      itemKey: 'Channel segment type',
      labelText: 'Channel segment type',
      itemType: FormTypes.ItemType.Display,
      displayText: parseWaveformChannelType(arrivalTimeFM.channel.name)
    },
    {
      itemKey: 'Author',
      labelText: 'Author',
      itemType: FormTypes.ItemType.Display,
      displayText: 'TBD'
    },
    {
      itemKey: 'Deleted',
      labelText: 'Deleted',
      itemType: FormTypes.ItemType.Display,
      displayText: SignalDetectionTypes.Util.getCurrentHypothesis(
        signalDetection.signalDetectionHypotheses
      )?.deleted
        ? 'True'
        : 'False'
    }
  );

  if (filterDefinitionByFilterDefinitionUsage) {
    formItems.push(
      {
        itemKey: 'Detection Filter',
        labelText: 'Detection Filter',
        itemType: FormTypes.ItemType.Display,
        displayText:
          filterDefinitionByFilterDefinitionUsage[FilterTypes.FilterDefinitionUsage.DETECTION].name
      },
      {
        itemKey: 'Onset Filter',
        labelText: 'Onset Filter',
        itemType: FormTypes.ItemType.Display,
        displayText:
          filterDefinitionByFilterDefinitionUsage[FilterTypes.FilterDefinitionUsage.ONSET].name
      },
      {
        itemKey: 'FK Filter',
        labelText: 'FK Filter',
        itemType: FormTypes.ItemType.Display,
        displayText:
          filterDefinitionByFilterDefinitionUsage[FilterTypes.FilterDefinitionUsage.FK].name
      }
    );
  }

  return formItems;
}
