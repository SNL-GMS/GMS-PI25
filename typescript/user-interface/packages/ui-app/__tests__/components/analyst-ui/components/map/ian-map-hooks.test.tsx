import { processingAnalystConfigurationData } from '@gms/common-model/__tests__/__data__';
import type { AnalystWaveformTypes, AppState } from '@gms/ui-state';
import { getStore, useAppDispatch, useAppSelector } from '@gms/ui-state';
import { appState } from '@gms/ui-state/__tests__/test-util';
import { act, renderHook } from '@testing-library/react-hooks';
import cloneDeep from 'lodash/cloneDeep';
import React from 'react';
import { Provider } from 'react-redux';

import type { MapContextMenusCallbacks } from '~analyst-ui/components/map/map-context-menus';

import {
  useEventOnClickHandler,
  useHideShowContextMenuState,
  useIANMapEventRightClickHandler,
  useIsMapSyncedToWaveformZoom,
  useMapNonPreferredEventData,
  useMapPreferredEventData,
  useSdOnClickHandler,
  useSdOnRightClickHandler,
  useSetIsMapSyncedToWaveformZoom,
  useStationOnClickHandler,
  useStationOnRightClickHandler
} from '../../../../../src/ts/components/analyst-ui/components/map/ian-map-hooks';
import { useQueryStateResult } from '../../../../__data__/test-util-data';
import { renderReduxHook } from '../../../../utils/render-hook-util';
import { eventResultsWithRejected } from '../events/event-data-types';

const processingAnalystConfigurationQuery = cloneDeep(useQueryStateResult);
processingAnalystConfigurationQuery.data = processingAnalystConfigurationData;

// eslint-disable-next-line @typescript-eslint/no-magic-numbers
jest.setTimeout(60000);

jest.mock('@gms/ui-state', () => {
  const actualRedux = jest.requireActual('@gms/ui-state');
  const mockDispatchFunc = jest.fn();
  const mockDispatch = () => mockDispatchFunc;
  const mockUseAppDispatch = jest.fn(mockDispatch);
  return {
    ...actualRedux,
    useAppDispatch: mockUseAppDispatch,
    useAppSelector: jest.fn((stateFunc: (state: AppState) => any) => {
      const stationsVisibility: AnalystWaveformTypes.StationVisibilityChangesDictionary = {};
      stationsVisibility.name = {
        visibility: true,
        stationName: 'station-name',
        isStationExpanded: false
      };
      const state: AppState = appState;
      const range = { startTimeSecs: 100, endTimeSecs: 200 };
      state.app.workflow.timeRange = range;
      state.app.workflow.openIntervalName = 'AL1';
      state.app.waveform.viewableInterval = range;
      state.app.waveform.stationsVisibility = stationsVisibility;
      state.app.common.selectedStationIds = ['station-name'];
      state.app.analyst.selectedSdIds = ['sd-name'];
      return stateFunc(state);
    }),
    useGetProcessingAnalystConfigurationQuery: jest.fn(() => processingAnalystConfigurationQuery),
    useGetEvents: jest.fn(() => ({
      ...eventResultsWithRejected
    })),
    useWorkflowQuery: jest.fn(() => ({
      isSuccess: true,
      data: { stages: [{ name: 'Auto Network' }, { name: 'AL1' }] }
    }))
  };
});

const store = getStore();

const MOCK_TIME = 1606818240000;
global.Date.now = jest.fn(() => MOCK_TIME);

describe('ui map', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  test('is defined', () => {
    expect(useHideShowContextMenuState).toBeDefined();
    expect(useStationOnClickHandler).toBeDefined();

    expect(useIsMapSyncedToWaveformZoom).toBeDefined();
    expect(useSetIsMapSyncedToWaveformZoom).toBeDefined();

    expect(useMapPreferredEventData).toBeDefined();
    expect(useMapNonPreferredEventData).toBeDefined();
    expect(useEventOnClickHandler).toBeDefined();

    expect(useIANMapEventRightClickHandler).toBeDefined();
    expect(useSdOnRightClickHandler).toBeDefined();
    expect(useStationOnRightClickHandler).toBeDefined();
  });

  it('call useSetIsMapSyncedToWaveformZoom', () => {
    useSetIsMapSyncedToWaveformZoom(true);
    expect(useAppDispatch).toHaveBeenCalledTimes(1);
  });

  it('call useMapPreferredEventData', () => {
    const result = renderReduxHook(store, () => useMapPreferredEventData());
    expect(result).toMatchSnapshot();
  });

  it('call useMapNonPreferredEventData', () => {
    const result = renderReduxHook(store, () => useMapNonPreferredEventData());
    expect(result).toMatchSnapshot();
  });

  it('call useStationOnClickHandler', () => {
    const mockDispatch = useAppDispatch();
    const { result } = renderHook(() => useStationOnClickHandler());
    act(() => {
      const entity: any = {
        id: 'station-name',
        properties: {
          type: {
            getValue: jest.fn(() => 'Station')
          }
        }
      };
      const onClick = result.current(entity);
      onClick();
    });
    expect(useAppSelector).toHaveBeenCalledTimes(1);
    expect(useAppDispatch).toHaveBeenCalledTimes(2);
    expect(mockDispatch).toHaveBeenCalledTimes(1);
    expect(mockDispatch).toHaveBeenCalledWith(expect.any(Function));
    act(() => {
      const entity: any = {
        id: 'station-wrong-name',
        properties: {
          type: {
            getValue: jest.fn(() => 'Station')
          }
        }
      };
      const onClick = result.current(entity);
      onClick();
    });
    expect(useAppSelector).toHaveBeenCalledTimes(1);
    expect(useAppDispatch).toHaveBeenCalledTimes(2);
    expect(mockDispatch).toHaveBeenCalledTimes(2);
    expect(mockDispatch).toHaveBeenCalledWith(expect.any(Function));
  });

  it('call useEventOnClickHandler', () => {
    const mockDispatch = useAppDispatch();
    const { result } = renderHook(() => useEventOnClickHandler());
    act(() => {
      const entity: any = {
        id: 'event-name',
        properties: {
          type: {
            getValue: jest.fn(() => 'Event location')
          }
        }
      };
      const onClick = result.current(entity);
      onClick();
    });
    expect(useAppSelector).toHaveBeenCalledTimes(1);
    expect(useAppDispatch).toHaveBeenCalledTimes(2);
    expect(mockDispatch).toHaveBeenCalledTimes(1);
    act(() => {
      const entity: any = {
        id: 'event-wrong-name',
        properties: {
          type: {
            getValue: jest.fn(() => 'Event location')
          }
        }
      };
      const onClick = result.current(entity);
      onClick();
    });
    expect(useAppSelector).toHaveBeenCalledTimes(1);
    expect(useAppDispatch).toHaveBeenCalledTimes(2);
    expect(mockDispatch).toHaveBeenCalledTimes(2);
    act(() => {
      const entity: any = {
        id: 'station',
        properties: {
          type: {
            getValue: jest.fn(() => 'Station')
          }
        }
      };
      const onClick = result.current(entity);
      onClick();
    });
    expect(useAppSelector).toHaveBeenCalledTimes(1);
    expect(useAppDispatch).toHaveBeenCalledTimes(2);
    expect(mockDispatch).toHaveBeenCalledTimes(2);
  });

  it('call useSdOnClickHandler', () => {
    const mockDispatch = useAppDispatch();
    const { result } = renderHook(() => useSdOnClickHandler());
    act(() => {
      const entity: any = {
        id: 'sd-name',
        properties: {
          type: {
            getValue: jest.fn(() => 'Signal detection')
          }
        }
      };
      const onClick = result.current(entity);
      onClick();
    });
    expect(useAppSelector).toHaveBeenCalledTimes(1);
    expect(useAppDispatch).toHaveBeenCalledTimes(2);
    expect(mockDispatch).toHaveBeenCalledTimes(1);
    act(() => {
      const entity: any = {
        id: 'sd-wrong-name',
        properties: {
          type: {
            getValue: jest.fn(() => 'Signal detection')
          }
        }
      };
      const onClick = result.current(entity);
      onClick();
    });
    expect(useAppSelector).toHaveBeenCalledTimes(1);
    expect(useAppDispatch).toHaveBeenCalledTimes(2);
    expect(mockDispatch).toHaveBeenCalledTimes(2);
  });

  describe('Map right-click handler hooks', () => {
    beforeEach(() => jest.clearAllMocks());

    function TestReduxWrapper({ children }: { children: React.ReactNode }) {
      return <Provider store={store}>{children}</Provider>;
    }

    const cb: MapContextMenusCallbacks = {
      eventContextMenuCb: jest.fn(),
      eventDetailsCb: jest.fn(),
      signalDetectionContextMenuCb: jest.fn(),
      signalDetectionDetailsCb: jest.fn(),
      stationContextMenuCb: jest.fn(),
      stationDetailsCb: jest.fn(),
      mapContextMenuCb: jest.fn()
    };

    const movementEvent: any = {
      position: {
        x: 3,
        y: 8
      }
    };

    const selectedEntityEvent: any = {
      properties: {
        id: {
          getValue: jest.fn(() => 'mockEventId')
        },
        event: {
          getValue: jest.fn(() => {
            return {
              event: {
                id: 'mockEventId',
                time: 0,
                latitudeDegrees: 45,
                longitudeDegrees: 45,
                depthKm: 10,
                status: 'Not Started'
              }
            };
          })
        },
        getValue: jest.fn(() => {
          return {
            event: {
              id: 'mockEventId',
              time: 0,
              latitudeDegrees: 45,
              longitudeDegrees: 45,
              depthKm: 10,
              status: 'Not Started'
            }
          };
        })
      }
    };

    const latitude = 25;
    const longitude = -71;

    it('useIANMapEventRightClickHandler opens the event context menu', () => {
      const mockSetEventId = jest.fn();

      // Get right-click handler function from hook
      const { result } = renderHook(() => useIANMapEventRightClickHandler(mockSetEventId, cb), {
        wrapper: TestReduxWrapper
      });

      result.current(movementEvent, selectedEntityEvent, latitude, longitude);

      expect(cb.eventContextMenuCb).toHaveBeenCalledTimes(1);
    });

    it('useStationOnRightClickHandler opens the station context menu', () => {
      // Get right-click handler function from hook
      const { result } = renderHook(() => useStationOnRightClickHandler(cb), {
        wrapper: TestReduxWrapper
      });

      result.current(movementEvent, selectedEntityEvent, latitude, longitude);

      expect(cb.stationContextMenuCb).toHaveBeenCalledTimes(1);
    });

    it('useSdOnRightClickHandler opens the signal detection context menu', () => {
      // Get right-click handler function from hook
      const { result } = renderHook(() => useSdOnRightClickHandler(cb), {
        wrapper: TestReduxWrapper
      });

      result.current(movementEvent, selectedEntityEvent, latitude, longitude);

      expect(cb.signalDetectionContextMenuCb).toHaveBeenCalledTimes(1);
    });
  });
});
