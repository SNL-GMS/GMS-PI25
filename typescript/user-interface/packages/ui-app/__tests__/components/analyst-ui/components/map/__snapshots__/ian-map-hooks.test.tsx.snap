// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui map call useMapNonPreferredEventData 1`] = `[]`;

exports[`ui map call useMapPreferredEventData 1`] = `
[
  {
    "activeAnalysts": [],
    "confidenceSemiMajorAxis": undefined,
    "confidenceSemiMajorAxisTrend": undefined,
    "confidenceSemiMinorAxis": undefined,
    "conflict": false,
    "coverageSemiMajorAxis": undefined,
    "coverageSemiMajorAxisTrend": undefined,
    "coverageSemiMinorAxis": undefined,
    "deleted": false,
    "depthKm": {
      "uncertainty": undefined,
      "value": 3.3,
    },
    "eventFilterOptions": [
      "After",
    ],
    "geoOverlappingEvents": 2,
    "id": "eventID",
    "isActionTarget": false,
    "isOpen": false,
    "isUnqualifiedActionTarget": false,
    "latitudeDegrees": 1.1,
    "longitudeDegrees": 2.2,
    "magnitudeMb": 1.2,
    "magnitudeMl": undefined,
    "magnitudeMs": undefined,
    "numberAssociated": 1,
    "numberDefining": 0,
    "observationsStandardDeviation": undefined,
    "preferred": true,
    "region": "TBD",
    "rejected": false,
    "status": undefined,
    "time": {
      "uncertainty": undefined,
      "value": 3600,
    },
    "unsavedChanges": false,
  },
  {
    "activeAnalysts": [],
    "confidenceSemiMajorAxis": undefined,
    "confidenceSemiMajorAxisTrend": undefined,
    "confidenceSemiMinorAxis": undefined,
    "conflict": false,
    "coverageSemiMajorAxis": undefined,
    "coverageSemiMajorAxisTrend": undefined,
    "coverageSemiMinorAxis": undefined,
    "deleted": false,
    "depthKm": {
      "uncertainty": undefined,
      "value": 3.3,
    },
    "eventFilterOptions": [
      "After",
    ],
    "geoOverlappingEvents": 2,
    "id": "eventID",
    "isActionTarget": false,
    "isOpen": false,
    "isUnqualifiedActionTarget": false,
    "latitudeDegrees": 1.1,
    "longitudeDegrees": 2.2,
    "magnitudeMb": 1.2,
    "magnitudeMl": undefined,
    "magnitudeMs": undefined,
    "numberAssociated": 1,
    "numberDefining": 0,
    "observationsStandardDeviation": undefined,
    "preferred": true,
    "region": "TBD",
    "rejected": false,
    "status": undefined,
    "time": {
      "uncertainty": undefined,
      "value": 3600,
    },
    "unsavedChanges": false,
  },
]
`;
